contrib/docs/configuration.md

| Option<br />(ENV setting)                           |                                   | Desecrption                                                  | Default                     |
| --------------------------------------------------- | --------------------------------- | ------------------------------------------------------------ | --------------------------- |
|                                                     |                                   | **General service settings**                                 |                             |
| debug<br />IP2L_DEBUG                               |                                   | (boolean) run in debugging mode, outputing informational and trace level log messages to stdout. | false                       |
| installation_path<br />IP2L_INSTALLATION_PATH       |                                   | (string) the path / location the service is installed,       | .                           |
| jsonlogs<br />IP2L_JSONLOGS                         |                                   | (not currently in use) output logs in a structured form      | false                       |
| loglevel<br />IP2L_LOG_LEVEL                        |                                   | (not currently in use) specify the level or severity of logging output. | debug                       |
|                                                     |                                   |                                                              |                             |
| ip2api                                              |                                   | **IP2Location API settings**                                 |                             |
| ip2api.key<br />|IP2L_IP2API_KEY                 |                                       | ip2location.io API key - sign-up for a free account at http://www.ip2location.io/ |                             |
| ip2api.max_errors<br />IP2L_IP2API_MAX_ERRORS     |                                     | (int) the number of API errors before the service terminates itself. <br />Set to 0 to disable (not recommended). | 5                           |
| ip2api.plan<br />IP2L_IP2API_PLAN      |                                               | (not currently in use)                                       |                             |
| ip2api.url<br />IP2L_IP2API_URL       |                                                 | (url) the URL to the IP2Location API                         | https://api.ip2location.io/ |
| ip2api.use_counters<br />IP2L_IP2API_USE_COUNTERS   |                                   | (boolean) if enabled by setting to true, the service will limit the number of calls made to the IP2Location.io API service to remain within the specified limits. If monthly is specified, counters will reset on a monthly basis with the number of calls remaining proratated over the number of remaining days until they are reset. |                             |
| ip2api.max_monthly<br />IP2L_IP2API_MAX_MONTHLY     |                                   | (int) Specify the limit for the total number of calls per month. At the beginning of the month, this number will be divided by the average nmber of days in a month (365/12) and daily limits will be applied to requests. Monthly counts are based on calendar months and are reset at midnight, 1st of each month. | 30,000                      |
| ip2api.max_daily<br />IP2L_IP2API_MAX_DAILY         |                                  | (int) Specify a lower bound or limit per day. Calls to the API will be blocked once this limit is reached, if set and if lower than the prorated monthly calls per day. Daily counts are reset at midnight every 24 hours. | 1,000                       |
| ip2api.max_perhour<br />IP2L_IP2API_MAX_PERHOUR     |                                   | (int) Similarly, if set this limit will apply per hour, if lower than both the number of prorated calls per day or per month. Hourly counts are reset on the hour - e.g. xx:00. |                             |
| ip2api.counters_file<br />IP2L_IP2API_COUNTERS_FILE |                                   | (string) the path and filename to maintain counts. If this file is missing or deleted, all counters will be reset. | ./counters.yaml             |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   | ** Service related settings **                               |                             |
| service.bind_host<br /><br />|IP2L_SERVICE_BIND_HOST                                    | (string) the IP address on the host to bind to. if bound to 127.0.0.1 the service will only be accessible to the host unless using a reverse proxy to allow communication with the pfSense host. May be set to 127.0.0.1 if runing on the pfSense host. | 0.0.0.0                     |
| service.bind_port<br />|IP2L_SERVICE_BIND_PORT                                          | (string) TCP port number to serve HTTP API requests requests and HTTP pages. | 9999                        |
| service.healthcheck<br />|IP2L_SERVICE_HEALTHCHECK                                      | (string) GET endpoint path for heath check calls             | /health                     |
| service.ingest_logs<br />|IP2L_SERVICE_INGEST_LOGS                                      | (string) POST endpoint path for the ingestion of pfSense filter.log entries | /api/filterlogs             |
| service.ip2l_results<br />|IP2L_SERVICE_IP2L_RESULTS                                    | (string) GET endpoint path for the querying and retrieval of result sets based on the supply of the resultsid as a query parameter | /api/results                |
| service.ip_requests<br />|IP2L_SERVICE_IP_REQUESTS                                      | (string) GET endpoint path for requesting details of a single IP address, supplied as a query parameter | /api/ip2location            |
| service.ip2geomap<br />|IP2L_SERVICE_IP2GEOMAP                                          | (string) endpoint to serve HTML pages for interacting with the service, including configuration, monitoring, querying and export of data | /index.html                 |
| service.use_ssl<br />|IP2L_SERVICE_USE_SSL                                              | (boolean) configures the API endpoints to use SSL/TLS, if set to true the path to a valid signed certificate and private key must be supplied. | false                       |
| service.ssl_cert<br />|IP2L_SERVICE_SSL_CERT                                            | (string) full path and filename of the SSL certificate       | (null)                      |
| service.ssl_key<br />         |       IP2L_SERVICE_SSL_KEY                              | (string) full path and filename of the SSL certificate's private key | (null)                      |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   | ** Redis cache settings **                                   |                             |
| use_redis                                           | IP2L_USE_REDIS                    | (boolean) utilise Redis cache - currently requried           | true                        |
| redis.ip2location.host                              | IP2L_REDIS_IP2LOCATION_HOST       | (string) hostname or IP address of the Redis server or cluster, the server or cluster may be shared with other applications | 127.0.0.1                   |
| redis.ip2location.port                              | IP2L_REDIS_IP2LOCATION_PORT       | (string) hostname or IP address of the Redis server or cluster | 6379                        |
| redis.ip2location.auth                              | IP2L_REDIS_IP2LOCATION_AUTH       | (string) authentication / username                           | ip2location                 |
| redis.ip2location.pass                              | IP2L_REDIS_IP2LOCATION_PASS       | (string) password                                            | password                    |
| redis.ip2location.db                                | IP2L_REDIS_IP2LOCATION_DB         | (int) database must be available for exlcusive for use by the service and must not shared with another applicaiton | 1                           |
| redis.ip2location.max.age                           | IP2L_REDIS_IP2LOCATION_MAX_AGE    | (int) as hours, the maximum age or TTL of items in the cache before they expire and are removed. Set to zero to disable aging. | 12                          |
| redis.ip2location.max.number                        | IP2L_REDIS_IP2LOCATION_MAX_NUMBER | (int) the maximum number of entries to hold at any one time. Both the age and the number of items will determine the life of an entry, with the first threshold reached applied - i.e. if age is set to 2 hours, and the number of items is set to 1,000 if 999 items are reached in the first hour, the entry will be removed at 1,000 rather than when it reaches 2 hours of age. Set to zero to disable limits. | 0                           |
| redis.pfsense.host                                  | IP2L_REDIS_PFSENSE_HOST           | (string) hostname or IP address of the Redis server or cluster, the server or cluster may be shared with other applications | 127.0.0.1                   |
| redis.pfsense.port                                  | IP2L_REDIS_PFSENSE_PORT           | (string) hostname or IP address of the Redis server or cluster | 6379                        |
| redis.pfsense.auth                                  | IP2L_REDIS_PFSENSE_AUTH           | (string) authentication / username                           | ip2location                 |
| redis.pfsense.pass                                  | IP2L_REDIS_PFSENSE_PASS           | (string) password                                            | password                    |
| redis.pfsense.db                                    | IP2L_REDIS_PFSENSE_DB             | (int) database assigned for the caching of enriched pfSense log entries / history to retain. | 2                           |
| redis.pfsense.max.age                               | IP2L_REDIS_PFSENSE_MAX_AGE        | (int) as hours, the maximum age or TTL of items in the cache before they expire and are removed. Set to zero to disable aging. recommend that the limits are more aggressive for this cache vs. the ip2location cache. | 2                           |
| redis.pfsense.max.number                            | IP2L_REDIS_PFSENSE_MAX_NUMBER     | (int) the maximum number of entries to hold at any one time. Both the age and the number of items will determine the life of an entry, with the first threshold reached applied - i.e. if age is set to 2 hours, and the number of items is set to 1,000 if 999 items are reached in the first hour, the entry will be removed at 1,000 rather than when it reaches 2 hours of age. Set to zero to disable limits. | 0                           |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   |                                                              |                             |
|                                                     |                                   |                                                              |                             |




backend/config/config.go
package config

import (
	"fmt"
	"os"
	"time"

	"github.com/jpmchia/ip2location-pfsense/util"

	"github.com/spf13/viper"
)

type Provider interface {
	ConfigFileUsed() string
	Get(key string) interface{}
	GetBool(key string) bool
	GetDuration(key string) time.Duration
	GetFloat64(key string) float64
	GetInt(key string) int
	GetInt64(key string) int64
	GetSizeInBytes(key string) uint
	GetString(key string) string
	GetStringMap(key string) map[string]interface{}
	GetStringMapString(key string) map[string]string
	GetStringMapStringSlice(key string) map[string][]string
	GetStringSlice(key string) []string
	GetTime(key string) time.Time
	InConfig(key string) bool
	IsSet(key string) bool
}

var Config Options

const appName string = "github.com/jpmchia/ip2location-pfsense/backend"
const appNamePrefix string = "ip2location-pfsense"

var CfgFile string = "config.yaml"
var defaultConfig *viper.Viper

// Default initialiser for the applicaiton's configuration
func init() {
	util.Log("[config] Initialising configuration")

	Configure()
}

func Configure() {
	// Load the default configuration
	defaultConfig = initViperConfig(appNamePrefix)

	// Set and load additional configuration locations
	setConfigLocations(CfgFile)

	// Read the configuration file
	err := defaultConfig.ReadInConfig()
	util.HandleError(err, "[config] Unable to read configuration:\n")

	// Unmarshal the configuration into the Config struct
	err = defaultConfig.Unmarshal(&Config)
	util.HandleError(err, "[config] Unable to unmarshal configuration:\n")
}

// Config returns a default config provider
func ConfigProvider() *viper.Viper {
	return defaultConfig
}

func GetConfiguration() Options {
	err := defaultConfig.Unmarshal(&Config)
	util.HandleError(err, "[config] Unable to unmarshal configuration:\n")
	return Config
}

func SetValue(key string, value interface{}) {
	defaultConfig.Set(key, value)
}

// SetConfigFile sets the config file to use and reloads the configuraton
func SetConfigFile(file string) {
	setConfigLocations(file)
	err := defaultConfig.ReadInConfig()
	util.HandleError(err, "[config] Unable to read configuration:\n")
}

// Initialises viper with default values
func initViperConfig(appName string) *viper.Viper {
	v := viper.New()
	v.SetEnvPrefix(appName)
	v.AutomaticEnv()

	// global defaults
	v.SetDefault("jsonlogs", false)
	v.SetDefault("loglevel", "debug")
	v.SetDefault("use_redis", true)
	v.SetDefault("installation_path", "/opt/ip2location-pfsense")
	v.SetDefault("redis.ip2location.host", "127.0.0.1")
	v.SetDefault("redis.ip2location.port", "6379")
	v.SetDefault("redis.ip2location.db", 1)
	v.SetDefault("redis.ip2location.auth", "ip2location")
	v.SetDefault("redis.ip2location.pass", "password")
	v.SetDefault("redis.pfsense.host", "127.0.0.1")
	v.SetDefault("redis.pfsense.port", "6379")
	v.SetDefault("redis.pfsense.db", 2)
	v.SetDefault("redis.pfsense.auth", "ip2location")
	v.SetDefault("redis.pfsense.pass", "password")
	v.SetDefault("redis.watchlist.host", "127.0.0.1")
	v.SetDefault("redis.watchlist.port", "6379")
	v.SetDefault("redis.watchlist.db", 3)
	v.SetDefault("redis.watchlist.auth", "ip2location")
	v.SetDefault("redis.watchlist.pass", "password")
	v.SetDefault("ip2api.url", "https://api.ip2location.io/")
	v.SetDefault("ip2api.key", "")
	v.SetDefault("ip2api.plan", "Free")
	v.SetDefault("ip2api.max_errors", 5)
	v.SetDefault("ip2api.source", "IP2Location-pfSense")
	v.SetDefault("service.bind_host", "127.0.0.1")
	v.SetDefault("service.bind_port", "9999")
	v.SetDefault("service.allow_hosts", "*")
	v.SetDefault("service.use_ssl", false)
	v.SetDefault("service.ssl_cert", "cert.pem")
	v.SetDefault("service.ssl_key", "cert.key")
	v.SetDefault("service.ingest_logs", "/api/filterlog")
	v.SetDefault("service.ip2l_results", "/api/results")
	v.SetDefault("service.ip2geomap", "/index.html")
	v.SetDefault("service.healthcheck", "/health")
	v.SetDefault("service.ip_requests", "/api/ip2location")
	v.SetDefault("use_cache", true)
	v.SetDefault("debug", false)

	return v
}

// SetConfigLocations sets the locations to search for the configuration file
func setConfigLocations(file string) {
	util.LogDebug("[config] Setting config locations")
	CfgFile = file
	// Use config file from the flag.
	defaultConfig.SetConfigType("yaml")
	defaultConfig.SetConfigFile(CfgFile)
	defaultConfig.AddConfigPath("./")
	defaultConfig.AddConfigPath(fmt.Sprintf("/etc/%s", appName))
	defaultConfig.AddConfigPath(fmt.Sprintf("/usr/local/%s", appName))
	defaultConfig.AddConfigPath(fmt.Sprintf("/opt/%s", appName))
	home, err := os.UserHomeDir()
	util.HandleError(err, "Unable to determine user's home directory")
	defaultConfig.AddConfigPath(fmt.Sprintf("%s/.%s", home, appName))
	defaultConfig.AddConfigPath(fmt.Sprintf("%s/.config/%s", home, appName))
	defaultConfig.AddConfigPath(home)
}

// PrintConfig prints the configuration to stdout
func printConfig(v *viper.Viper) {
	for _, k := range v.AllKeys() {
		fmt.Printf("%s = %v\n", k, v.Get(k))
	}
}

// WriteConfigValue writes a value to the configuration file
func WriteConfigValue(key string, value any) {
	defaultConfig.Set(key, value)

	err := defaultConfig.WriteConfig()
	util.HandleFatalError("Unable to write configuration:\n", err.Error())
}

// ShowConfig prints the configuration to stdout
func ShowConfig() {
	fmt.Println("Configuration:")
	fmt.Println("-------------")
	fmt.Printf("Configuration file: %s\n", defaultConfig.ConfigFileUsed())
	fmt.Println("-------------")

	printConfig(defaultConfig)

	os.Exit(0)
}

// CreateConfigFile creates a new configuration file with the default values and exits,
// if no filename is specified, the default filename is used
func CreateConfigFile(args []string) {
	if len(args) == 0 {
		fmt.Printf("No filename specified. Using the default file: %s", CfgFile)
	}
	if len(args) > 1 {
		fmt.Printf("Too many arguments specified. Using the default file: %s", CfgFile)
	}
	if len(args) == 1 {
		CfgFile = args[0]
	}

	util.LogDebug("Creating configuration file: %s", CfgFile)

	// If a config file is found, read it in.
	err := defaultConfig.SafeWriteConfigAs(CfgFile)
	util.HandleFatalError(err, "Unable to write configuration:\n")

	fmt.Printf("Configuration file created: %s", CfgFile)
}
